File: JustClient/Remap.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

/**
 * Script to remap intermediary names to yarn names in Java source files.
 * This converts class_XXX, method_XXX, field_XXX to their proper yarn names.
 */
public class Remap {
    // Path to the yarn mappings file
    static final String MAPPINGS_FILE = "C:\\Users\\stust\\.gradle\\caches\\fabric-loom\\1.21.4\\net.fabricmc.yarn.1_21_4.1.21.4+build.8-v2\\mappings.tiny";

    // Source directory
    static final String SRC_DIR = "C:\\Users\\stust\\eclipse-workspace\\ExonClient-Source\\src\\main\\java";

    static Map<String, String> classMap = new HashMap<>();
    static Map<String, String> methodMap = new HashMap<>();
    static Map<String, String> fieldMap = new HashMap<>();

    public static void main(String[] args) throws Exception {
        System.out.println("Loading mappings...");
        parseMappings();
        System.out.println("Loaded " + classMap.size() + " class mappings, " +
                          methodMap.size() + " method mappings, " +
                          fieldMap.size() + " field mappings");

        // Find all Java files
        List<Path> javaFiles = new ArrayList<>();
        Files.walk(Paths.get(SRC_DIR))
             .filter(p -> p.toString().endsWith(".java"))
             .forEach(javaFiles::add);

        System.out.println("Found " + javaFiles.size() + " Java files");

        // Remap each file
        int changed = 0;
        for (Path filepath : javaFiles) {
            if (remapFile(filepath)) {
                System.out.println("Remapped: " + filepath);
                changed++;
            }
        }

        System.out.println("\nDone! Changed " + changed + " files.");
    }

    static void parseMappings() throws Exception {
        try (BufferedReader reader = new BufferedReader(new FileReader(MAPPINGS_FILE))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.isEmpty() || line.startsWith("tiny")) continue;

                String[] parts = line.split("\t");

                if (line.startsWith("c\t") && parts.length >= 4) {
                    // Class mapping: c official intermediary named
                    String intermediary = parts[2]; // e.g., net/minecraft/class_310
                    String named = parts[3]; // e.g., net/minecraft/client/MinecraftClient

                    // Extract simple class name
                    String intSimple = intermediary.substring(intermediary.lastIndexOf('/') + 1);
                    String namedSimple = named.substring(named.lastIndexOf('/') + 1);

                    classMap.put(intSimple, namedSimple);
                    classMap.put(intermediary.replace('/', '.'), named.replace('/', '.'));

                } else if (line.startsWith("\tf\t") && parts.length >= 5) {
                    // Field mapping
                    String intermediary = parts[3];
                    String named = parts[4];
                    if (intermediary.startsWith("field_")) {
                        fieldMap.put(intermediary, named);
                    }

                } else if (line.startsWith("\tm\t") && parts.length >= 5) {
                    // Method mapping
                    String intermediary = parts[3];
                    String named = parts[4];
                    if (intermediary.startsWith("method_")) {
                        methodMap.put(intermediary, named);
                    }
                }
            }
        }
    }

    static boolean remapFile(Path filepath) throws Exception {
        String content = Files.readString(filepath);
        String original = content;

        // Replace class names (both in imports and in code)
        for (Map.Entry<String, String> entry : classMap.entrySet()) {
            String intermediary = entry.getKey();
            String named = entry.getValue();
            if (intermediary.contains("class_")) {
                content = content.replace(intermediary, named);
            }
        }

        // Replace method names
        for (Map.Entry<String, String> entry : methodMap.entrySet()) {
            String intermediary = entry.getKey();
            String named = entry.getValue();
            content = content.replaceAll("\\b" + Pattern.quote(intermediary) + "\\b", named);
        }

        // Replace field names
        for (Map.Entry<String, String> entry : fieldMap.entrySet()) {
            String intermediary = entry.getKey();
            String named = entry.getValue();
            content = content.replaceAll("\\b" + Pattern.quote(intermediary) + "\\b", named);
        }

        if (!content.equals(original)) {
            Files.writeString(filepath, content);
            return true;
        }
        return false;
    }
}

File: JustClient/remap.py
#!/usr/bin/env python3
"""
Script to remap intermediary names to yarn names in Java source files.
This converts class_XXX, method_XXX, field_XXX to their proper yarn names.
"""

import os
import re
import sys

# Path to the yarn mappings file
MAPPINGS_FILE = r"C:\Users\stust\.gradle\caches\fabric-loom\1.21.4\net.fabricmc.yarn.1_21_4.1.21.4+build.8-v2\mappings.tiny"

# Source directory
SRC_DIR = r"C:\Users\stust\eclipse-workspace\ExonClient-Source\src\main\java"

def parse_mappings(mappings_file):
    """Parse the tiny mappings file and return dictionaries for classes, methods, and fields."""
    class_map = {}  # intermediary -> yarn
    method_map = {}  # intermediary -> yarn
    field_map = {}  # intermediary -> yarn

    current_class_intermediary = None
    current_class_named = None

    with open(mappings_file, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.rstrip('\n')
            if not line or line.startswith('tiny'):
                continue

            parts = line.split('\t')

            if line.startswith('c\t'):
                # Class mapping: c official intermediary named
                if len(parts) >= 4:
                    intermediary = parts[2]  # e.g., net/minecraft/class_310
                    named = parts[3]  # e.g., net/minecraft/client/MinecraftClient

                    # Extract simple class name
                    int_simple = intermediary.split('/')[-1]
                    named_simple = named.split('/')[-1]

                    class_map[int_simple] = named_simple
                    class_map[intermediary.replace('/', '.')] = named.replace('/', '.')

                    current_class_intermediary = int_simple
                    current_class_named = named_simple

            elif line.startswith('\tf\t'):
                # Field mapping: \tf descriptor official intermediary named
                if len(parts) >= 5:
                    intermediary = parts[3]
                    named = parts[4]
                    if intermediary.startswith('field_'):
                        field_map[intermediary] = named

            elif line.startswith('\tm\t'):
                # Method mapping: \tm descriptor official intermediary named
                if len(parts) >= 5:
                    intermediary = parts[3]
                    named = parts[4]
                    if intermediary.startswith('method_'):
                        method_map[intermediary] = named

    return class_map, method_map, field_map

def remap_file(filepath, class_map, method_map, field_map):
    """Remap a single Java source file."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original = content

    # Replace class names (both in imports and in code)
    for intermediary, named in class_map.items():
        if 'class_' in intermediary:
            # For full package paths like net.minecraft.class_310
            content = content.replace(intermediary, named)

    # Replace method names
    for intermediary, named in method_map.items():
        # Replace method calls like .method_123(
        content = re.sub(r'\b' + re.escape(intermediary) + r'\b', named, content)

    # Replace field names
    for intermediary, named in field_map.items():
        content = re.sub(r'\b' + re.escape(intermediary) + r'\b', named, content)

    if content != original:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return True
    return False

def main():
    print("Loading mappings...")
    class_map, method_map, field_map = parse_mappings(MAPPINGS_FILE)
    print(f"Loaded {len(class_map)} class mappings, {len(method_map)} method mappings, {len(field_map)} field mappings")

    # Find all Java files
    java_files = []
    for root, dirs, files in os.walk(SRC_DIR):
        for file in files:
            if file.endswith('.java'):
                java_files.append(os.path.join(root, file))

    print(f"Found {len(java_files)} Java files")

    # Remap each file
    changed = 0
    for filepath in java_files:
        if remap_file(filepath, class_map, method_map, field_map):
            print(f"Remapped: {filepath}")
            changed += 1

    print(f"\nDone! Changed {changed} files.")

if __name__ == '__main__':
    main()
